ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <stdbool.h>
+#include <math.h>
+#include <pthread.h>
+#include <limits.h>
+#include <sys/wait.h>
+
+#include <caml/fail.h>
+#include <caml/memory.h>
+#include <caml/unixsupport.h>
+#include <caml/signals.h>
+
+#include "../helper/vfork_helper.h"
+
+#define FOREACH_LIST(name, list) \
+    for(value name = (list); name != Val_emptylist; name = Field(name, 1))
+
+// Create thread reducing stack usage to a minimum to reduce memory usage.
+// Returns error number (like pthread_create).
+static int create_thread_minstack(pthread_t *th, void *(*proc)(void *), void *arg);
+
+static inline void
+reap_pid(pid_t pid)
+{
+    int status;
+    while (waitpid(pid, &status, 0) < 0 && errno == EINTR)
+        continue;
+}
+
+static void *
+thread_proc_reap(void *arg)
+{
+    pid_t pid = (pid_t) (intptr_t) arg;
+
+    reap_pid(pid);
+
+    return NULL;
+}
+
+// Appends a string to *p_dest buffer.
+// It updates *p_dest to point after copied string.
+// Returns copied string.
+static char *
+append_string(char **p_dest, const char *s)
+{
+    char *const dest = *p_dest;
+    size_t const size = strlen(s) + 1;
+    memcpy(dest, s, size);
+    *p_dest = dest + size;
+    return dest;
+}
+
+static char**
+copy_string_list(value list)
+{
+    size_t strings_size = 0;
+    size_t list_size = 0;
+    char **res, **ptrs;
+    char *strings;
+
+    FOREACH_LIST(item, list) {
+        strings_size += strlen(String_val(Field(item, 0))) + 1;
```

Comment:
It's true. It's not exactly the same, although in this case it does not matter and can be replaced with the proposed function.

Reply:
`strlen` only works correctly if there are no null bytes in the OCaml string - which in general are legal in OCaml.

Reply:
It depends on the definition of correctness and what's your target. As said before, beside spending potentially (but unlikely) a bit more memory, there won't be no changes. These strings are meant to get to some system calls which are written in C. That is, they expect NUL-terminated strings. Ocaml are more of "NUL-extended" strings (not official terminology), in the sense that the length does not contain a NUL terminator character and the strings can contain NUL characters in the middle but a NUL character is always appended to the buffer to be easily passed and used by C code. So... if there are no NUL character in the middle, `strlen` will report correct C string length, while `caml_string_length` won't. As said before however we are passing strings meant for C code so they should not contain NUL character in the middle. That is, `strlen` and `caml_string_length` should report the same value. In any case, anything after the first NUL will be ignored. So functionally does not matter. Using `caml_string_length` will probably save some CPU cycles, avoiding scanning the strings.

To be picky, in this case strings should not have NUL characters embedded, and an error could be reported, but this will fall back to a `strlen` cost.

Note that the above code is very similar to different library (also in Ocaml implementation) so if you think that's wrong you are stating the same of Ocaml internal library or dune, for instance.